codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyTLV.java

```diff
@@ -0,0 +1,145 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.handler.codec.haproxy;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.DefaultByteBufHolder;
+
+import static io.netty.util.internal.ObjectUtil.*;
+
+/**
+ * A Type-Length Value (TLV vector) that can be added to the PROXY protocol
+ * to include additional information like SSL information.
+ *
+ * @see HAProxySSLTLV
+ */
+public class HAProxyTLV extends DefaultByteBufHolder {
+
+    private final Type type;
+    private final byte typeByteValue;
+
+    /**
+     * The registered types a TLV can have regarding the PROXY protocol 1.5 spec
+     */
+    public enum Type {
+        PP2_TYPE_ALPN,
+        PP2_TYPE_AUTHORITY,
+        PP2_TYPE_SSL,
+        PP2_TYPE_SSL_VERSION,
+        PP2_TYPE_SSL_CN,
+        PP2_TYPE_NETNS,
+        /**
+         * A TLV type that is not officially defined in the spec. May be used for nonstandard TLVs
+         */
+        OTHER;
+
+        /**
+         * Returns the the {@link Type} for a specific byte value as defined in the PROXY protocol 1.5 spec
+         * <p>
+         * If the byte value is not an official one, it will return {@link Type#OTHER}.
+         *
+         * @param byteValue the byte for a type
+         * @return the {@link Type} of a TLV
+         */
+        public static Type typeForByteValue(final byte byteValue) {
+            switch (byteValue) {
+            case 0x01:
+                return PP2_TYPE_ALPN;
+            case 0x02:
+                return PP2_TYPE_AUTHORITY;
+            case 0x20:
+                return PP2_TYPE_SSL;
+            case 0x21:
+                return PP2_TYPE_SSL_VERSION;
+            case 0x22:
+                return PP2_TYPE_SSL_CN;
+            case 0x30:
+                return PP2_TYPE_NETNS;
+            default:
+                return OTHER;
+            }
+        }
+    }
+
+    /**
+     * Creates a new HAProxyTLV
+     *
+     * @param type the {@link Type} of the TLV
+     * @param typeByteValue the byteValue of the TLV. This is especially important if non-standard TLVs are used
+     * @param content the raw content of the TLV
+     */
+    HAProxyTLV(final Type type, final byte typeByteValue, final ByteBuf content) {
+        super(content);
+        checkNotNull(type, "type");
+
+        this.type = type;
+        this.typeByteValue = typeByteValue;
+    }
+
+    /**
+     * Returns the {@link Type} of this TLV
+     */
+    public Type type() {
+        return type;
+    }
+
+    /**
+     * Returns the type of the TLV as byte
+     */
+    public byte typeByteValue() {
+        return typeByteValue;
+    }
+
+    @Override
+    public HAProxyTLV copy() {
+        return (HAProxyTLV) super.copy();
+    }
+
+    @Override
+    public HAProxyTLV duplicate() {
+        return (HAProxyTLV) super.duplicate();
```

Comment:
Related to the previous comment ... The typical way this is done in other classes is something like this:

```java
    @Override
    public HAProxyTLV duplicate() {
        return replace(content().duplicate());
    }

    @Override
    public HAProxyTLV replace(ByteBuf content) {
        return new HAProxyTLV(type, typeByteValue, content);
    }
```

Then all of the other methods which do the same thing (copy, retainedDuplicate, etc...) can be written the same way.